orttraining/orttraining/python/training/ortmodule/ort_triton/_decompose.py

```diff
@@ -0,0 +1,367 @@
+        name = self._get_unique_var_name(f"{node_name}_{op_type}_")
+        if outputs is None:
+            outputs = [f"{name}_out"]
+        for idx, output in enumerate(outputs):
+            if output is None:
+                outputs[idx] = f"{name}_out{idx}"
+        return helper.make_node(op_type, inputs, outputs, name, **kwargs), *outputs
+
+    def _get_dtype_and_shape(self, arg_name: str, **kwargs):
+        node_arg_infos = kwargs["node_arg_infos"]
+        arg_info = node_arg_infos[arg_name]
+        return arg_info.dtype, arg_info.shape
+
+    def _decompose_elementwise_precision(self, node: NodeProto, graph: GraphProto, **kwargs):
+        x = node.input[0]
+        dtype, _ = self._get_dtype_and_shape(x, **kwargs)
+        if dtype != TensorProto.FLOAT16 and dtype != TensorProto.BFLOAT16:
+            return [node]
+        node_name = node.name
+        y = node.output[0]
+        op_type = node.op_type
+        inputs = [input for input in node.input]
+        cast_nodes = []
+        for idx, input in enumerate(inputs):
+            dtype, _ = self._get_dtype_and_shape(input, **kwargs)
+            if dtype == TensorProto.FLOAT16 or dtype == TensorProto.BFLOAT16:
+                cast_node, cast_out = self._new_node(node_name, "Cast", [input], to=TensorProto.FLOAT)
+                cast_nodes.append(cast_node)
+                inputs[idx] = cast_out
+        op_node, op_out = self._new_node(node_name, op_type, inputs)
+        cast_node1, _ = self._new_node(node_name, "Cast", [op_out], outputs=[y], to=dtype)
+        return [*cast_nodes, op_node, cast_node1]
+
+    def Exp(self, node: NodeProto, graph: GraphProto, **kwargs):  # noqa: N802
+        return self._decompose_elementwise_precision(node, graph, **kwargs)
+
+    def Pow(self, node: NodeProto, graph: GraphProto, **kwargs):  # noqa: N802
+        return self._decompose_elementwise_precision(node, graph, **kwargs)
+
+    def Sqrt(self, node: NodeProto, graph: GraphProto, **kwargs):  # noqa: N802
+        return self._decompose_elementwise_precision(node, graph, **kwargs)
+
+    def LayerNormalization(self, node: NodeProto, graph: GraphProto, **kwargs):  # noqa: N802
+        node_name = node.name
+        x = node.input[0]
+        w = node.input[1]
+        b = node.input[2]
+        y = node.output[0]
+        mean = node.output[1] if len(node.output) > 1 and node.output[1] else None
+        inv_std_dev = node.output[2] if len(node.output) > 2 and node.output[2] else None
+        axis = get_attribute(node, "axis", -1)
+        epsilon = get_attribute(node, "epsilon", 1e-05)
+        xdtype, shape = self._get_dtype_and_shape(x, **kwargs)
+        wdtype, _ = self._get_dtype_and_shape(w, **kwargs)
+        is_x_half = xdtype == TensorProto.FLOAT16 or xdtype == TensorProto.BFLOAT16
+        is_w_half = wdtype == TensorProto.FLOAT16 or wdtype == TensorProto.BFLOAT16
+        if is_x_half or is_w_half:
```

Comment:
What about using the function body defined in onnx?

Reply:
If I correctly understand this part, this folder aims at more or less writing onnx graphs. This feature is implemented in many places (onns-script, ...). By experience, it takes times to maintain. To reduce the maintenance cost, I would create a very simple interface the triton feature needs. That way, the triton specific code can evolve independantly and won't be impacted if we decide to rewrite that code with another solution maintained somewhere else assuming the interface between the two is stable. I would only keep in this folder one file with this interface and put the current code in another folder.

Reply:
The code from onnx may not handle the half data type well. And not all the ONNX Op has corresponding function in Triton, end up we need to rewrite every single decompose logic instead of reuse any one from ONNX.

Reply:
This approach is similar to what tvm does. Maybe we could consider Triton as a new provider and organize the code same way. Enabling triton would be done by add "TritonExecutionProvider" to the list of providers.

Reply:
You are right on this. But currently this feature is for training only, and supports a small part of Ops. From the huggingface models it helps to improve the perf for training, but not sure about the case for inference. Making it a new EP requires lots of changes to the code base in inference (or common) side, we can do this once we are sure this is good for both cases.